hikari/applications.py

```diff
@@ -560,8 +567,104 @@ class AuthorizationInformation:
     expires_at: datetime.datetime = attr.ib(eq=True, hash=False, repr=True)
     """When the access token this data was retrieved with expires."""
 
-    scopes: typing.Sequence[str] = attr.ib(eq=True, hash=False, repr=True)
+    scopes: typing.Sequence[typing.Union[OAuth2Scope, str]] = attr.ib(eq=True, hash=False, repr=True)
     """A sequence of the scopes the current user has authorized the application for."""
 
     user: typing.Optional[users.User] = attr.ib(eq=True, hash=False, repr=True)
     """The user who has authorized this token if they included the `identify` scope."""
+
+
+@attr_extensions.with_copy
+@attr.s(eq=True, hash=True, init=True, kw_only=True, slots=True, weakref_slot=False)
+class PartialOAuth2Token:
+    """Model for partial OAuth2 token data returned by the API.
+
+    This will generally only be returned when by the client credentials OAuth2
+    flow.
+    """
+
+    access_token: str = attr.ib(eq=True, hash=True, repr=False)
+    """Access token issued by the authorization server."""
+
+    token_type: typing.Union[TokenType, str] = attr.ib(eq=False, hash=False, repr=True)
+    """Type of token issued by the authorization server."""
+
+    expires_in: datetime.timedelta = attr.ib(eq=False, hash=False, repr=True)
+    """Lifetime of this access token."""
+
+    scopes: typing.Sequence[typing.Union[OAuth2Scope, str]] = attr.ib(eq=False, hash=False, repr=True)
+    """Scopes the access token has access to."""
+
+    def __str__(self) -> str:
+        return self.access_token
+
+
+@attr_extensions.with_copy
+@attr.s(eq=True, hash=True, init=True, kw_only=True, slots=True, weakref_slot=False)
+class OAuth2AuthorizationToken(PartialOAuth2Token):
+    """Model for the OAuth2 token data returned by the authorization grant flow."""
+
+    refresh_token: int = attr.ib(eq=False, hash=False, repr=False)
+    """Refresh token used to obtain new access tokens with the same grant."""
+
+    webhook: typing.Optional[webhooks.Webhook] = attr.ib(eq=False, hash=False, repr=True)
+    """Object of the webhook that was created.
+
+    This will only be present if this token was authorized with the
+    `webhooks.incoming` scope, otherwise this will be `builtins.None`.
+    """
+
+    guild: typing.Optional[guilds.RESTGuild] = attr.ib(eq=False, hash=False, repr=True)
+    """Object of the guild the user was added to.
+
+    This will only be present if this token was authorized with the
+    `bot` scope, otherwise this will be `builtins.None`.
+    """
+
+
+@attr_extensions.with_copy
+@attr.s(eq=True, hash=True, init=True, kw_only=True, slots=True, weakref_slot=False)
+class OAuth2ImplicitToken(PartialOAuth2Token):
+    """Model for the OAuth2 token data returned by the implicit grant flow."""
+
+    state: typing.Optional[str] = attr.ib(eq=False, hash=False, repr=False)
+    """State parameter that was present in the authorization request if provided."""
```

Comment:
The problem now is that applications are now a different thing (eg, bots). I think it would be best to separate them

Reply:
Well most of the objects and enums in that module are only returned by oauth2 endpoints (pretty sure ApplicationFlags is the only exception since you get that on identify).

Reply:
Also fwiw an application in the context of Discord is an oauth2 application which can optionally have features like a bot user or "embedded" functionality etc etc attached to it.